packages/gestalt/src/Masonry/dynamicHeightsUtils.ts

```diff
@@ -0,0 +1,96 @@
+/**
+ * Util functions used to update positions when an item changes the height dynamically
+ */
+import { Cache } from './Cache';
+import { ColumnSpanConfig } from './multiColumnLayout';
+import { Position } from './types';
+
+function shouldExtendArea(area: { top: number; left: number; right: number }, position: Position) {
+  return position.left < area.left || position.left + position.width > area.right;
+}
+
+function isBelowArea(area: { top: number; left: number; right: number }, position: Position) {
+  return (
+    position.left < area.right &&
+    position.left + position.width > area.left &&
+    position.top > area.top
+  );
+}
+
+function recalcHeights<T>({
+  items,
+  changedItem,
+  newHeight,
+  positionStore,
+  measurementStore,
+  getColumnSpanConfig,
+}: {
+  items: ReadonlyArray<T>;
+  changedItem: T;
+  newHeight: number;
+  positionStore: Cache<T, Position>;
+  measurementStore: Cache<T, number>;
+  getColumnSpanConfig?: (item: T) => ColumnSpanConfig;
+}) {
+  const changedItemPosition = positionStore.get(changedItem);
+
+  if (!changedItemPosition || changedItemPosition.height === newHeight) {
+    return;
+  }
+
+  measurementStore.set(changedItem, newHeight);
+
+  const { top, left, width, height } = changedItemPosition;
+  const itemsWithPositions = items.filter((item) => item && positionStore.has(item));
+  const multiColumnItemsPositions = getColumnSpanConfig
+    ? itemsWithPositions
+        .filter((item) => getColumnSpanConfig(item) !== 1)
+        .map((item) => {
+          const position = positionStore.get(item);
+          return position ? { item, position } : undefined;
+        })
+        .filter((itemPosition) => !!itemPosition)
+        .sort((a, b) => a.position.top - b.position.top)
+    : [];
+
+  const areas = multiColumnItemsPositions.reduce(
```

Comment:
can you explain why areas needs to be an array vs just a left + right bound?

Reply:
This is because the affected areas are tied to the height, in certain heights only certai area applies fe just bellow the item but on other heights the areas can grow.